staging/src/k8s.io/cloud-provider/credentialconfig/registry.go

```diff
@@ -0,0 +1,101 @@
+package credentialconfig
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// Code taken from /pkg/credentialprovider/config.go
+
+// RegistryConfig represents the config file used by the docker CLI.
+// This config that represents the credentials that should be used
+// when pulling images from specific image repositories.
+type RegistryConfig map[string]RegistryConfigEntry
+
+// RegistryConfigEntry wraps a docker config as a entry
+type RegistryConfigEntry struct {
```

Comment:
@andyzhangx @nilo19 Continuing our discussion from https://github.com/kubernetes-sigs/cloud-provider-azure/issues/289, I was hoping this might work for the Azure external credential provider. This should be fairly close to `DockerConfigEntry`, sans `Provider` and a couple other details. Converting from `RegistryConfigEntry` to `DockerConfigEntry` is as simple as:

```go
func registryToDocker(registryConfig credentialconfig.RegistryConfig) credentialprovider.DockerConfig {
	dockerConfig := credentialprovider.DockerConfig{}
	for k, v := range registryConfig {
		dockerConfig[k] = credentialprovider.DockerConfigEntry{
			Username: v.Username,
			Password: v.Password,
			Email:    v.Email,
		}
	}
	return dockerConfig
}
```

@andrewsykim - feel free to weigh in if you have any thoughts, too!

Reply:
@DangerOnTheRanger thanks, lgtm. And what's the next action when this PR get merged? Do I need to implement similar functionalities in staging folder for Azure credential provider?

Reply:
Refactoring out the main functionality of the credential provider into a package in staging/ while leaving behind anything specific to pkg/credentialprovider should be sufficient. It needs to undergo some review, but the GCP credential provider code in this PR should provide a good example, with both pkg/credentialprovider/gcp and staging/k8s.io/legacy-cloud-providers/gce/gcpcredential.